File: mini_world.py
class Sun:
    is_day_time = True
    def change_day_time(self):
        self.is_day_time = not self.is_day_time
        print("Now it is day time") if self.is_day_time else print("Now it is night") 
        
class Tree:
    def make_air(self, sun : Sun):
        print("The tree is making air") if sun.is_day_time else print("The tree is not making air as it is night")
        
class Frog:
    x = 0
    y = 0
    is_on_the_grass = True
    def update_frog_position(self):
        if self.x > 3 and self.y > 3 and self.x < 5 and self.y < 5:
            self.is_on_the_grass = False
            return
        self.is_on_the_grass = True
    
    def move_right(self):
        self.x += 1
        self.update_frog_position()
    def move_left(self):
        self.x -= 1
        self.update_frog_position()
    def move_up(self):
        self.y += 1
        self.update_frog_position()
    def move_down(self):
        self.y -= 1
        self.update_frog_position()
    
    def is_sleeping(self, sun : Sun):
        return True if not sun.is_day_time else False
        
class Grass:
    area = 10
    def to_be_eaten(self, frog : Frog):
        if self.area <= 0:
            print("There is no grass. Wait till tommorow")
        elif frog.is_sleeping(sun):
            print("The frog can't eat the grass as it is sleeping. Call the sunrise to wake it up")
        elif not frog.is_on_the_grass:
            print("The frog can't eat the grass as it is on the lake. Move it to the grass")
        else:
            self.area -= 1
            print("The frog ate one m^2 area of grass")
    
    def add_area(self, sun : Sun):
        if sun.is_day_time:
            self.area += 1
            print("1 m^2 area of grass has been added")
        else:
            print("You can't add area in the night time")
    
sun = Sun()
tree = Tree()
grass = Grass()
frog = Frog()

sun.change_day_time()
tree.make_air(sun)
grass.to_be_eaten(frog)
sun.change_day_time()
for i in range(4):
    frog.move_right()
    frog.move_up()
grass.to_be_eaten(frog)
frog.move_left()
frog.move_left()
for i in range(11):
    grass.to_be_eaten(frog)
sun.change_day_time()
grass.add_area(sun)
sun.change_day_time()
grass.add_area(sun)
grass.to_be_eaten(frog)

File: mini_world.c
#include <stdio.h>
#include <stdbool.h>

struct Sun{
    char is_day_time;
};

struct Tree{};

struct Frog{
    int x;
    int y;
    char is_on_the_grass;
};

struct Grass{
    int area;
};

struct Mini_World{
    struct Sun sun;
    struct Tree tree;
    struct Frog frog;
    struct Grass grass;
};


void Sun(struct Mini_World* ptr, char n){
    (ptr->sun).is_day_time = n;
}

void change_day_time(struct Mini_World* ptr){
    (ptr->sun).is_day_time = !(ptr->sun).is_day_time;
    if((ptr->sun).is_day_time == 1){
        printf("Now it is day time\n");
    }else{
        printf("Now it is night\n");
    }
    return;
}

void make_air(struct Mini_World* ptr){
    if((ptr->sun).is_day_time == 1){
        printf("The tree is making air\n");
    }else{
        printf("The tree is not making air as it is night\n");
    }
    return;
}

void update_frog_position(struct Mini_World* ptr){
    if((ptr->frog).x > 3 && (ptr->frog).y > 3 && (ptr->frog).x < 5 && (ptr->frog).y < 5){
        (ptr->frog).is_on_the_grass = 0;
        return;
    }
    (ptr->frog).is_on_the_grass = 1;
}

void Frog(struct Mini_World* ptr, int x, int y){
    (ptr->frog).x = x;
    (ptr->frog).y = y;
    update_frog_position(ptr);
}

void move_right(struct Mini_World* ptr){
    ((ptr->frog).x)++;
    update_frog_position(ptr);
}

void move_left(struct Mini_World* ptr){
    ((ptr->frog).x)--;
    update_frog_position(ptr);
}

void move_up(struct Mini_World* ptr){
    ((ptr->frog).y)++;
    update_frog_position(ptr);
}

void move_down(struct Mini_World* ptr){
    ((ptr->frog).y)--;
    update_frog_position(ptr);
}

bool is_sleeping(struct Mini_World* ptr){
    return (ptr->sun).is_day_time == 0;
}

void Grass(struct Mini_World* ptr, int area){
    ptr->grass.area = area;
}

void to_be_eaten(struct Mini_World* ptr){
    if((ptr->grass).area <= 0){
        printf("There is no grass. Wait till tommorow\n");
        return;
    }else if(is_sleeping(ptr)){
        printf("The frog can't eat the grass as it is sleeping. Call the sunrise to wake it up\n");
        return;
    }else if(!(ptr->frog).is_on_the_grass){
        printf("The frog can't eat the grass as it is on the lake. Move it to the grass\n");
        return;
    }else{
        ((ptr->grass).area)--;
        printf("The frog ate one m^2 area of grass\n");
    }
}

void add_area(struct Mini_World* ptr){
    if((ptr->sun).is_day_time){
        (ptr->grass).area++;
        printf("1 m^2 area of grass has been added\n");
    }else{
        printf("You can't add area in the night time\n");
    }
}

int main()
{
    struct Mini_World mini_world;
    Sun(&mini_world, 1);
    Frog(&mini_world, 4, 4);
    Grass(&mini_world, 1);
    
    change_day_time(&mini_world);
    change_day_time(&mini_world);
    make_air(&mini_world);
    change_day_time(&mini_world);
    make_air(&mini_world);
    to_be_eaten(&mini_world);
    change_day_time(&mini_world);
    to_be_eaten(&mini_world);
    move_left(&mini_world);
    to_be_eaten(&mini_world);
    to_be_eaten(&mini_world);
    add_area(&mini_world);
    to_be_eaten(&mini_world);
    
    

    return 0;
}
